controller/daemon.go

```diff
@@ -69,8 +76,38 @@ func (ctrl *Daemon) Run(ctx context.Context) error {
 		return err
 	}
 
-	exitBufLen := 2 // api & run tasks exit
-	exitCh := make(chan error, exitBufLen)
+	var rm *registration.SelfRegistrationManager
+	if *conf.Consul.SelfRegistration.Enabled {
+		// Expect one more long-running goroutine
+		exitBufLen++
+		exitCh = make(chan error, exitBufLen)
+
+		// Configure Consul client if not already
+		if ctrl.consulClient == nil {
+			c, err := client.NewConsulClient(conf.Consul, client.ConsulDefaultMaxRetry)
+			if err != nil {
+				ctrl.logger.Error("error setting up Consul client", "error", err)
+				return err
+			}
+			ctrl.consulClient = c
+		}
+
+		// Configure and start self-registration manager
+		rm = registration.NewSelfRegistrationManager(
+			&registration.SelfRegistrationManagerConfig{
+				ID:               *conf.ID,
+				Port:             *conf.Port,
+				TLSEnabled:       (conf.TLS != nil && *conf.TLS.Enabled),
+				SelfRegistration: conf.Consul.SelfRegistration,
+			},
+			ctrl.consulClient)
+		go func() {
```

Comment:
For better readability, I suggest spacing out code blocks specially if indentations and parenthesis are involved :)